cedar-policy-validator/src/typecheck.rs

```diff
@@ -2021,18 +2019,37 @@ impl<'a> Typechecker<'a> {
                         }
                     }
                 }
-                // Unspecified entities will be detected by a different part of the validator.
-                // Still return `TypecheckFail` so that typechecking is not considered successful.
-                Some(EntityType::Unspecified) => TypecheckAnswer::fail(
-                    ExprBuilder::with_data(Some(Type::primitive_boolean()))
-                        .with_same_source_loc(in_expr)
-                        .is_in(lhs_expr, rhs_expr),
-                ),
+                Some(EntityType::Unspecified) => {
+                    // It's perfectly valid for `principal` or `resource` to be `EntityType::Unspecified`
```

Comment:
I don't think this is a behavior change, because the case where we return `Type::singleton_boolean(false)` is actually unreachable -- we would have already returned at line 1762 and never even reached this function.

For the question about the Dafny spec, I believe the Dafny spec returns singleton-false for this case as well (`unspecified in RHS` where `RHS` is a single specified entity literal or set of specified entity literals) at lines 417-418 at your link, and never reaches the line 439 which you pointed to.

Perhaps both the Rust and Dafny need additional comments to clarify which cases actually reach particular lines.  It took me a long time to puzzle these out as well.  I feel that neither codebase is particularly well organized in how it typechecks `in`.

Reply:
It still seems incorrect to ignore the failure of `TypecheckFail` though.

Reply:
I think changing the behavior many layers above to not ignore `TypecheckFail` would be a great idea but is well beyond the scope of this PR.  Perhaps we could do that first though?  If we're right that this is unreachable, then it shouldn't matter whether we return Fail or Success, at least until some refactor makes this code reachable again

Reply:
I agree with Kesha's assessment of the un-reachability of both branches. The `then` statement could only be reached if an entity literal had an unspecified entity type, but that can't happen. The else branch is more subtle, but Craig seems to be correct that the unspecified-in-specifed case rules out that branch.